### PR TITLE
Orca: Implement MainCallBacks in TorchRunner

### PR DESCRIPTION
**Pull request by leonardozcm**
## Description

MainCallback is a one-of-a-kind callback that contains hook functions:
    - `on_iter_forward`
    - `on_iter_backward`
    - `on_lr_adjust`
These methods are somewhat special, because only one special MainCallback should be allowed to implement these methods among all callbacks, otherwise there will propagate forward and backward twice.

### CODE REVIEW

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 1]
```diff
@@ -206,7 +207,7 @@ def setup_operator(self, training_models):
             self.dist_backend = TorchDistBackend()
 
     def train_epochs(self, data_creator, epochs=1, batch_size=32, profile=False,
-                     info=None, wrap_dataloader=None, callbacks=None,
+                     info=None, wrap_dataloader=None, callbacks=[],
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:210]: seems in Python, using a mutable list as default value is not recommended?
Still keep it None and change to an empty list in the function.

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 2]
```diff
@@ -253,12 +254,14 @@ def train_epochs(self, data_creator, epochs=1, batch_size=32, profile=False,
...
             val_steps = None
 
-        if callbacks is not None:
-            for callback in callbacks:
-                callback.set_model(self.given_models)
-                if hasattr(callback, "set_trainer"):
-                    callback.set_trainer(self)
-                callback.on_train_begin()
+        # Check uniqueness of the MainCallback
+        make_only_mainCallback(callbacks)
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:258]: What about check it in the estimator? The error message thrown by the main process may be easier to find?

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks/maincallback.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    # TODO: Refactor scheduler update logic in TorchRunner
+    def on_lr_adjust(self, runner):
+        if runner.lr_scheduler is not None:
+            runner.lr_scheduler.step()
+
+    def on_train_forward(self, runner):
+        self.on_iter_forward(runner)
+
+    def on_val_forward(self, runner):
+        self.on_iter_forward(runner)
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks/maincallback.py:71]: Make it clearer in the document whether or when the users need to override these two methods?

Comment (leonardozcm) [reply python/orca/src/bigdl/orca/learn/pytorch/callbacks/maincallback.py:71]: sure, and other hooks in base.py will be added in anther pr.

File: python/orca/src/bigdl/orca/learn/pytorch/callbacks/maincallback.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    for i in range(len(callbacks)):
+        if isinstance(callbacks[i], MainCallback):
+            _num_MCB += 1
+            # MainCallback should be called at the highest priority
+            callbacks[0], callbacks[i] = callbacks[i], callbacks[0]
+
+    if _num_MCB == 0:
+        callbacks.insert(0, MainCallback())
+    elif _num_MCB > 1:
+        invalidInputError(False, f"Expect only one or no MainCallback"
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/callbacks/maincallback.py:31]: Expect at most one ...

File: python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py [hunk 5]
```diff
@@ -240,7 +240,8 @@ def fit(self,
         :param label_cols: label column names if data is Spark DataFrame.
         :param validation_data: validation data. Validation data type should be the same
                as train data.
-        :param callbacks: A list for all callbacks.
+        :param callbacks: A list for all callbacks. Note that only one special MainCallback
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py:243]: remove `special`?

File: python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py [hunk 6]
```diff
@@ -240,7 +240,8 @@ def fit(self,
...
+               should be allowed to implement these methods among all callbacks.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py:244]: should be -> is
remove `to implement these methods`?

File: python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py [hunk 7]
```diff
@@ -515,6 +522,8 @@ def evaluate(self,
                for validate.
         :param feature_cols: feature column names if train data is Spark DataFrame.
         :param label_cols: label column names if train data is Spark DataFrame.
+        :param callbacks: A list for all callbacks. Note that only one special MainCallback
+               should be allowed to implement these methods among all callbacks.
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py:526]: same as above

File: python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py [hunk 8]
```diff
@@ -492,22 +492,11 @@ def _train_batch(self, batch, batch_info=None, callbacks=None):
...
 
         # Compute gradients in a backward pass.
-        with self.timers.record("grad"):
-            self.optimizer.zero_grad()
-            self.loss.backward()
-
-        # Call step of optimizer to update model params.
-        with self.timers.record("apply"):
-            self.optimizer.step()
+        with self.timers.record("grad&apply"):
```
Comment (hkvision) [inline python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py:498]: change to bwd to align with fwd?

Comment (leonardozcm) [reply python/orca/src/bigdl/orca/learn/pytorch/torch_runner.py]: all above have been fixed.

**Review by hkvision — APPROVED** [review]

**Pull request merged**